## Step 0 — Import Libraries

In [1]:
import numpy as np

np.set_printoptions(precision=4, suppress=True)
print("NumPy version:", np.__version__)

NumPy version: 2.1.3


## Step 1 — Create My Personal Matrix

Student ID = **0432220005101083**

- `d1` = last digit of student ID = **3**
- `d2` = second last digit of student ID = **8**

In [2]:
student_id = "0432220005101083"

d1 = int(student_id[-1])   # last digit
d2 = int(student_id[-2])   # second last digit

print("d1 (last digit)        =", d1)
print("d2 (second last digit) =", d2)

A = np.array([[d1 + 2, d2 + 1],
              [2 * d1, d2 + 2]], dtype=float)

print("\nMatrix A =")
print(A)

d1 (last digit)        = 3
d2 (second last digit) = 8

Matrix A =
[[ 5.  9.]
 [ 6. 10.]]


## Step 2 — Analyze Matrix A

Compute: Shape, Determinant, Rank, Eigenvalues, and Inverse (only if determinant != 0).

In [3]:
def analyze(M, name):
    print("=" * 45)
    print("Analysis of Matrix", name)
    print("=" * 45)
    print(M)
    print("-" * 45)

    print("Shape       :", M.shape)

    det = np.linalg.det(M)
    print("Determinant :", round(det, 4))

    rank = np.linalg.matrix_rank(M)
    print("Rank        :", rank)

    eigvals = np.linalg.eigvals(M)
    print("Eigenvalues :", np.round(eigvals, 4))

    if not np.isclose(det, 0):
        inv = np.linalg.inv(M)
        print("Inverse     :")
        print(np.round(inv, 4))
    else:
        inv = None
        print("Inverse     : Not possible (determinant = 0, matrix is singular)")

    print("Condition No:", round(np.linalg.cond(M), 4))
    print()
    return det, rank, eigvals, inv


detA, rankA, eigA, invA = analyze(A, "A")

Analysis of Matrix A
[[ 5.  9.]
 [ 6. 10.]]
---------------------------------------------
Shape       : (2, 2)
Determinant : -4.0
Rank        : 2
Eigenvalues : [-0.2621 15.2621]
Inverse     :
[[-2.5   2.25]
 [ 1.5  -1.25]]
Condition No: 60.4835



## Step 3 — Value Change Experiment

I create a new matrix **B** by changing exactly **one** element of A.

> **Element changed:** `A[0][0]` (row 1, column 1) — I added **+1** to it.
>
> So `A[0][0] = 5` becomes `B[0][0] = 6`. Every other element stays the same.

In [4]:
B = A.copy()
B[0, 0] = B[0, 0] + 1     # <-- the single element I changed

print("Original A =")
print(A)
print("\nModified B =")
print(B)
print("\nChanged element: position [0][0] ->", A[0, 0], "became", B[0, 0])

Original A =
[[ 5.  9.]
 [ 6. 10.]]

Modified B =
[[ 6.  9.]
 [ 6. 10.]]

Changed element: position [0][0] -> 5.0 became 6.0


## Step 4 — Re-analyze Matrix B

In [5]:
detB, rankB, eigB, invB = analyze(B, "B")

Analysis of Matrix B
[[ 6.  9.]
 [ 6. 10.]]
---------------------------------------------
Shape       : (2, 2)
Determinant : 6.0
Rank        : 2
Eigenvalues : [ 0.3842 15.6158]
Inverse     :
[[ 1.6667 -1.5   ]
 [-1.      1.    ]]
Condition No: 42.1429



## Step 5 — Side by Side Comparison

In [6]:
print(f"{'Property':<15}{'Matrix A':<28}{'Matrix B':<28}")
print("-" * 71)
print(f"{'Shape':<15}{str(A.shape):<28}{str(B.shape):<28}")
print(f"{'Determinant':<15}{round(detA,4):<28}{round(detB,4):<28}")
print(f"{'Rank':<15}{rankA:<28}{rankB:<28}")
print(f"{'Eigenvalues':<15}{str(np.round(eigA,4)):<28}{str(np.round(eigB,4)):<28}")
print(f"{'Trace':<15}{round(np.trace(A),4):<28}{round(np.trace(B),4):<28}")
print(f"{'Condition No':<15}{round(np.linalg.cond(A),4):<28}{round(np.linalg.cond(B),4):<28}")
print("-" * 71)

print("\nInverse of A:")
print(np.round(invA, 4))
print("\nInverse of B:")
print(np.round(invB, 4))

print("\nVerification (A @ A_inv should be Identity):")
print(np.round(A @ invA, 4))
print("\nVerification (B @ B_inv should be Identity):")
print(np.round(B @ invB, 4))

Property       Matrix A                    Matrix B                    
-----------------------------------------------------------------------
Shape          (2, 2)                      (2, 2)                      
Determinant    -4.0                        6.0                         
Rank           2                           2                           
Eigenvalues    [-0.2621 15.2621]           [ 0.3842 15.6158]           
Trace          15.0                        16.0                        
Condition No   60.4835                     42.1429                     
-----------------------------------------------------------------------

Inverse of A:
[[-2.5   2.25]
 [ 1.5  -1.25]]

Inverse of B:
[[ 1.6667 -1.5   ]
 [-1.      1.    ]]

Verification (A @ A_inv should be Identity):
[[1. 0.]
 [0. 1.]]

Verification (B @ B_inv should be Identity):
[[1. 0.]
 [0. 1.]]


## Step 6 — Comparison & Explanation

### 1. How did the determinant change and why?

The determinant changed from **-4** for matrix A to **+6** for matrix B — a jump of exactly **+10**, and it also flipped its sign.

This happens because for a 2x2 matrix the determinant is `ad - bc`. I added +1 to the element `a` (position [0][0]), so the new determinant becomes `(a+1)d - bc = (ad - bc) + d`. Here `d = A[1][1] = 10`, which is exactly why the determinant increased by 10. So a change in one element does not shift the determinant randomly — it shifts it by the cofactor of that element. The sign flip means the two row vectors changed their orientation: A was very close to singular (|det| = 4 is small compared to the size of the entries), and pushing one element by just 1 was enough to cross zero and come out on the other side.

### 2. Did the rank change? Explain.

**No.** The rank stayed **2** for both A and B.

Rank tells us how many linearly independent rows (or columns) the matrix has. Both A and B have a non-zero determinant (-4 and +6), and for a 2x2 matrix any non-zero determinant automatically means full rank, i.e. rank = 2. The rank would only have dropped to 1 if the change had made the determinant exactly 0, which would require the two rows to become multiples of each other. Since that did not happen, both matrices still span the whole 2D space and the rank is unchanged.

### 3. How did the eigenvalues respond to the value change?

For **A** the eigenvalues are approximately **15.2621** and **-0.2621**, and for **B** they are approximately **15.6158** and **0.3842**.

Two things happened. First, the sum of the eigenvalues (the trace) grew from 15 to 16, exactly the +1 I added, because the element I changed sits on the main diagonal. Second, the product of the eigenvalues (the determinant) moved from -4 to +6, so the small eigenvalue flipped from negative to positive. The large eigenvalue barely moved (about +0.35), but the small one changed sign completely. This shows that the dominant eigenvalue is quite stable against a small perturbation, while the small eigenvalue is very sensitive — a tiny change in the data can completely change its behaviour.

### 4. Is B easier or harder to invert than A? Why?

**B is easier to invert.**

Both matrices are invertible, but the quality of the inversion is different. The determinant of A is only -4 while the determinant of B is +6, and since the inverse is computed as `(1/det) * adj(M)`, a smaller determinant means we divide by a smaller number and the entries of the inverse blow up. The condition number makes this clear: A has a condition number of about **60.48** whereas B is about **42.14**. A higher condition number means the matrix is closer to singular ("ill-conditioned"), so any small rounding or measurement error in the input gets amplified much more in the output. Since B has the larger determinant and the lower condition number, B is numerically more stable and therefore easier and safer to invert.

